test_scalar_cube_load

marqh opened the pull request:
new test to check that a scalar data variable in a netcdf file will be loaded into a cube

addressing long standing bug reports, including #1427

cpelley commented:
Thanks @marqh

marqh commented on lib/iris/fileformats/cf.py:
```diff
@@ -487,8 +487,8 @@ def identify(cls, variables, ignore=None, target=None, warn=True, monotonic=Fals
                 continue
-            # Restrict to one-dimensional with name as dimension OR zero-dimensional scalar
-            if not ((nc_var.ndim == 1 and nc_var_name in nc_var.dimensions) or (nc_var.ndim == 0)):
+            # Restrict to one-dimensional with name as dimension
```
A zero dimensional scalar variable can never be a netcdf coordinate variable.  The definitions within the CF Conventions and the netCDF user guide are explicit on this.  
A 'coordinate variable' is one dimensional, with the same name as the dimension it is defining.
A scalar variable is not defined with respect to a dimension, it is a scalar.

marqh commented:
I am getting test failures unrelated to my change, e.g.
https://travis-ci.org/SciTools/iris/jobs/292568227#L5125
```
 File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/dask/utils.py", line 688, in __call__

return getattr(obj, self.method)(*args, **kwargs)

AttributeError: 'NetCDFDataProxy' object has no attribute 'reshape'
```

djkirkham commented:
Looks like an issue around the new masked array functionality in Dask. `dask.array.from_array` now has a keyword argument `asarray` which, when false, prevents Dask converting each chunk to an `ndarray`. This is intended to allow masked arrays to be passed through. For non-scalar arrays the chunks are the results of calls to `__getitem__` on the underlying object, but for scalar arrays there is no `__getitem__` call, so in this particular case the `NetCDFDataProxy` object is being returned by dask at the lowest level, and a `reshape()` call is being made on it (by Dask), which fails. I've raised this issue on the Dask repo (https://github.com/dask/dask/issues/2823).

pelson commented:
@djkirkham: would you mind helping to add a *unit test* that covers the issue identified with dask?

djkirkham commented:
I've added a fix workaround and unit tests in https://github.com/SciTools/iris/pull/2878

pelson commented:
@marqh, would you mind rebasing this and adding a what's new? I think the tests should go green what that done, and I'd be happy to merge for the ``2.0.0rc1``.

marqh commented:
> would you mind rebasing this and adding a what's new? I think the tests should go green what that done, and I'd be happy to merge for the 2.0.0rc1.

@pelson  I have rebased and added a what's new sentence to the 2.0.rst directly

awaiting test results ... ...

pelson commented:
Thanks @marqh.
